dijet/tasks/jersf.py

```diff
@@ -0,0 +1,294 @@
+    def reduce_histogram(self, histogram, processes, shifts):
+        import hist
+
+        def flatten_nested_list(nested_list):
+            return [item for sublist in nested_list for item in sublist]
+
+        # transform into lists if necessary
+        processes = law.util.make_list(processes)
+        # categories = law.util.make_list(categories)
+        shifts = law.util.make_list(shifts)
+
+        # get all leaf categories
+        # category_insts = list(map(self.config_inst.get_category, categories))
+        # leaf_category_insts = set(flatten_nested_list([
+        #     category_inst.get_leaf_categories() or [category_inst]
+        #     for category_inst in category_insts
+        # ]))
+
+        # get all sub processes
+        process_insts = list(map(self.config_inst.get_process, processes))
+        sub_process_insts = set(flatten_nested_list([
+            [sub for sub, _, _ in proc.walk_processes(include_self=True)]
+            for proc in process_insts
+        ]))
+
+        # get all shift instances
+        shift_insts = [self.config_inst.get_shift(shift) for shift in shifts]
+
+        # work on a copy
+        h = histogram.copy()
+
+        # axis selections
+        h = h[{
+            "process": [
+                hist.loc(p.id)
+                for p in sub_process_insts
+                if p.id in h.axes["process"]
+            ],
+            "shift": [
+                hist.loc(s.id)
+                for s in shift_insts
+                if s.id in h.axes["shift"]
+            ],
+        }]
+
+        # axis reductions
+        h = h[{"process": sum, "shift": sum}]
+
+        return h
+
+
+class alpha(HistogramsBaseTask):
+    def output(self) -> dict[law.FileSystemTarget]:
+        output = {
+            "dummy": self.target("dummy.pickle"),
+            # NOTE: do we want to keep plots in a nested directory or in a sibling directory?
+            "asym": self.target("plots/asym", dir=True, optional=True),
+            "alpha": self.target("plots/alpha", dir=True, optional=True),
+            "jer": self.target("plots/jer", dir=True, optional=True),
+        }
+        return output
+
+    def run(self):
+        h_data_all = []
+        h_mc_all = []
+        for dataset in self.datasets:
+            isMC = "qcd" in dataset
+            for variable in self.variables:
+                h_in = self.load_histogram(dataset, variable)
+
+                # Store all hists in a list to sum over after reading
+                if isMC:
+                    h_mc_all.append(h_in)
+                else:
+                    h_data_all.append(h_in)
+
+        h_data = sum(h_data_all)
+        h_mc = sum(h_mc_all)
+
+        alpha_bins = [1, 2, 3, 4, 5, 6]
+        eta_bins = h_data.axes["probejet_abseta"].edges
+        pt_bins = h_data.axes["dijets_pt_avg"].edges
+        pt_centers = h_data.axes["dijets_pt_avg"].centers
+        asym_centers = h_data.axes["dijets_asymmetry"].centers
```

Comment:
It looks like this task requires specific variables. Does one need to pass these variabels via the `--variables` flag? What happens if they are omitted or the incorrect ones are supplied?

Maybe the `variables` parameter should be overridden with a class-level variable containing the (hardcoded) variable names.

Reply:
The variables your a talking about in the four lines of codes exist per definition. The histograms we call in this task already contain the (pt/eta/elpha) binning of interest and hence can be just called here.

The `--variables` called in `load_histogram` is equivalent to the one when creating the hist. Meaning if it does not exist, it can not be called, but it is not a unique feature of this task but based on law itself.

Reply:
We discussed this a bit privately, but just to summarize. The issue is that the variables might not exist in the input histogram depending on what is provided on the command line via `--variables`. At the moment, these need to correspond exactly to the variables used in the `run` method, or the `AlphaExtrapolation` task will fail. Since the alpha extrapolation anyway always relies on exactly these three/four variables, it is probably preferable to hard-code the variables rather than having them as a user-supplied parameter.

Nevertheless, the issue is probably better addressed in a separate PR. For now, we can merge this, with the caveat that specific `--variables` need to be supplied in the correct order when calling the `AlphaExtrapolation` task.